### PR TITLE
Improving Llama Vision Models Cross-Attention Mask Implementation for Lower Prefill Latency

### PR DESCRIPTION
**Pull request by ssanjayTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/23383

### Problem description
Currently the way cross-attention masks work in llama-vision models is that we allocate a single Tensor of shape (1, 1, max_generation_length, num_vision_tokens) for each request in a batch during prefill which is stored in cpu memory. This introduces significant memory overhead (around 3GB per user for Llama-3.2-11B-Vision-Instruct) which causes hangs for batch-size 16 and also increases the time to first token (TTFT). 

## Dynamic Cross Attention Mask Allocation

Instead of allocating for a single cross-attention mask across max generation length we create two tensors: a `prefill-cross-attention_mask` and a `decode_cross_attention_mask`. The `prefill_cross_attention_mask` is allocated for all prefill tokens and ids specified by the vision masks. The `decode_cross_attention_mask` is used for future tokens generated by the decoder as specified by the vision_masks. This significantly lowers prefill time and resolves the prefill hangs at higher batch sizes. For an image of size 512x512 with isl=128 and osl=128, the TTFT drops from 3000ms to 2000ms.

## Removed unnecessary copying of vision_tokens from tt-device to cpu and back

This was being done to pad vision_tokens with 0 values. however this can be done using ttnn, so we do not need to perform this roundtrip. This reduces TTFT by around 400 ms using tt-inference-server benchmarks.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by yieldthought — APPROVED** [review]
Nice 👍

**Comment by mtairum** [comment]
Great improvement. Please make sure that the T3K demo pipeline passes before merging.: https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml

Ideally, run all T3K pipelines: https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml

**Review by cglagovichTT — APPROVED** [review]
Nice work!

**Comment by ssanjayTT** [comment]
> Great improvement. Please make sure that the T3K demo pipeline passes before merging.: https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml
> 
> Ideally, run all T3K pipelines: https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml

How do I run this pipeline? I'm not on a branch, I'm working off a fork.

**Comment by smehtaTT** [comment]
@tt-rkim - can you help check this?

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -556,15 +565,17 @@ def prefill_forward(
...
     def decode_forward(
         self,
         start_pos,
         tokens,
-        cross_attention_masks,
-        full_text_row_masked_out_mask,
+        prefill_cross_attention_masks,
+        prefill_full_text_row_masked_out_mask,
+        decode_cross_attention_masks,
+        decode_full_text_row_masked_out_mask,
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Since the prefill and decode generator interfaces are modified, please don't merge this until the corresponding vLLM PR has been reviewed and is ready to be merged.

**Comment by tt-rkim** [comment]
You cannot run T3K demo or model perf tests from a fork. You can only run APC. 

A models team member can copy your branch for you and run the tests for you: https://github.com/tenstorrent/tt-metal/actions/workflows/mirror-fork-branch.yaml

But if you're a TT internal developer, I'm not sure if you need to develop from a fork unless you just want to practice this flow and enforce these boundaries.

**Comment by ssanjayTT** [comment]
I have the vllm fix here https://github.com/tenstorrent/vllm/pull/119

**Comment by ssanjayTT** [comment]
Here are the results of the demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/15882053362
Here are the results of all t3k tests: https://github.com/tenstorrent/tt-metal/actions/runs/15879396163

I think the demo tests are flaky. I've run the t3000_llama3_vision_tests and the performance is higher than expected. So I'm pretty sure the ci is randomly failing. Here's a link to the log of my test run on the n300: https://gist.github.com/ssanjayTT/7dce152ac33e3b021d9761e6f18ecef2

**Comment by ssanjayTT** [comment]
Update on the demo tests: There was a ci issue which I believe might have been resolved and the vision test now passes https://github.com/tenstorrent/tt-metal/actions/runs/15933218645/job/44947931687 on ci 🎉. Our prefill_tok/s has increased from 6.5 Tok/s to 13.2 Tok/s 🥳 for the 11B-Vision model

**Comment by mtairum** [comment]
Nice! Seems all blockers are taken care off. 

The All post commit pipeline seems to be failing but, the issue is unrelated to this commit. My suggestion would be to rebase your branch, rerun those failing tests and the merge this 👍
